src/main/java/com/es/phoneshop/model/dao/ArrayListProductDao.java

```diff
@@ -46,9 +47,9 @@ public List<Product> findProducts() {
     public void save(Product product) {
         synchronized (lock) {
             try {
-                getProduct(product.getId()).update(product);
+                products.set(Math.toIntExact(getProduct(product.getId()).getId()), product);
             }
-            catch(ProductNotFoundException exception) { // if getProduct throws exception -> no such id in products or id == null
+            catch(RuntimeException exception) { // if getProduct throws exception -> no such id in products or id == null
```

Comment:
Don't use exceptions  like that. It's slowing down you application. You'll have exception each time you have to create a new product. All this info will be in the logs. It would be very missleading for other developers who doesn't know that you save method relies on exceptions.
I would refactor this method like below:
if(product.getId() != null) { // get product by id => update if exists or create a new one if null } else { // create a new one }